### PR TITLE
run tests with self compiled ddmd

### PR DESCRIPTION
**Pull request by MartinNowak**
- with all initial host compilers (dmd/gdc/ldc)

### CODE REVIEW

File: travis.sh [hunk 1]
```diff
@@ -4,13 +4,27 @@ set -exo pipefail
...
-make -j$N -C src -f posix.mak dmd.conf HOST_DMD=$DMD
 git clone --depth=1 https://github.com/D-Programming-Language/druntime.git ../druntime
 git clone --depth=1 https://github.com/D-Programming-Language/phobos.git ../phobos
+
+make -j$N -C src -f posix.mak HOST_DMD=$DMD all
+make -j$N -C src -f posix.mak HOST_DMD=$DMD dmd.conf
 make -j$N -C ../druntime -f posix.mak
 make -j$N -C ../phobos -f posix.mak
 
+if [ $SELF_COMPILE -eq 1 ]; then
```
Comment (MartinNowak) [inline travis.sh]: We could actually turn that into a loop and recompile a number of times.
```bash
while [ $SELF_COMPILE -gt 0 ]; do
    // rebuild...
    SELF_COMPILE=$(($SELF_COMPILE - 1))
done
```

Comment (MartinNowak) [inline travis.sh]: done

**Comment by ibuclaw** [comment]
Is the 3rd (last) ddmd built used to run unittests/testsuite?  Or are we time constrained by the travis build bots?

**Comment by MartinNowak** [comment]
> Is the 3rd (last) ddmd built used to run unittests/testsuite? Or are we time constrained by the travis build bots?

No it should work fine. It's the test suite that takes so much time.

**Comment by ibuclaw** [comment]
Ok. By the way do we have any cache available to keep first stage compiler tarballs in?

File: travis.sh (hunk 1 above)
Comment (yebblies) [inline travis.sh]: Wouldn't it be better to self-compile once then diff the binaries?  Compiling again with identical binaries seems like a waste of time...

**Comment by ibuclaw** [comment]
Auto-merge toggled on

**Comment by ibuclaw** [comment]
Auto-merge toggled off

**Comment by ibuclaw** [comment]
I received an email from @yebblies but I can't see it here. But I agree. :-)

File: travis.sh [hunk 2]
```diff
@@ -4,13 +4,28 @@ set -exo pipefail
...
+make -j$N -C src -f posix.mak HOST_DMD=$DMD dmd.conf
 make -j$N -C ../druntime -f posix.mak
 make -j$N -C ../phobos -f posix.mak
 
+while [ $SELF_COMPILE -gt 0 ]; do
+    # rebuild dmd using the just build dmd as host compiler
+    mv src/dmd src/host_dmd
+    make -j$N -C src -f posix.mak HOST_DMD=./host_dmd clean
+    make -j$N -C src -f posix.mak HOST_DMD=./host_dmd dmd.conf
+    make -j$N -C src -f posix.mak HOST_DMD=./host_dmd
```
Comment (ibuclaw) [inline travis.sh]: This is OK.

File: travis.sh [hunk 3]
```diff
@@ -4,13 +4,28 @@ set -exo pipefail
 
 N=2
 
-make -j$N -C src -f posix.mak HOST_DMD=$DMD
-make -j$N -C src -f posix.mak dmd.conf HOST_DMD=$DMD
 git clone --depth=1 https://github.com/D-Programming-Language/druntime.git ../druntime
 git clone --depth=1 https://github.com/D-Programming-Language/phobos.git ../phobos
+
+make -j$N -C src -f posix.mak HOST_DMD=$DMD all
+make -j$N -C src -f posix.mak HOST_DMD=$DMD dmd.conf
```
Comment (ibuclaw) [inline travis.sh]: First dmd.conf, then all?

**Comment by MartinNowak** [comment]
> Ok. By the way do we have any cache available to keep first stage compiler tarballs in?

Not really, there is a number of cache but none that fit the purpose.

> I received an email from @yebblies but I can't see it here. But I agree. :-)

I don't see the mail, but I can figure the content.
The same "optimization" is done on the auto-tester (see [here](https://github.com/braddr/at-client/blob/661e4cbaa45c11ffe84edb6fd0bda8c73ceb0186/src/do_test_dmd.sh#L28)) and I think we should apply it as well, b/c 6x0.5h testing is quite some load for a free service.

**Comment by MartinNowak** [comment]
> I don't see the mail, but I can figure the content.

Whoops, wrong PR.

**Comment by ibuclaw** [comment]
> I don't see the mail, but I can figure the content.

He referred to self compiling and diffing the binaries, rather than clean and remake phobos/druntime.

I've just ran the following on my computer:
```
# Stage 1
make -j8 -f posix.mak CC=g++-4.9 HOST_DMD=gdmd
mv dmd dmd.stage1
mv dmd.o dmd.stage1.o
make -j8 -f posix.mak CC=g++-4.9 HOST_DMD=gdmd clean

# Stage 2
make -j8 -f posix.mak CC=g++-4.9 HOST_DMD=./dmd.stage1
mv dmd dmd.stage2
mv dmd.o dmd.stage2.o
make -j8 -f posix.mak CC=g++-4.9 HOST_DMD=./dmd.stage1 clean

# Stage 3
make -j8 -f posix.mak CC=g++-4.9 HOST_DMD=./dmd.stage2
mv dmd dmd.stage3
mv dmd.o dmd.stage3.o
make -j8 -f posix.mak CC=g++-4.9 HOST_DMD=./dmd.stage2 clean

# Binary check
md5sum -b dmd.stage2.o
md5sum -b dmd.stage3.o
```

And the end result is:
```
75a6f00d5a3ddaa32e68ee7456c3eb92 *dmd.stage2.o
b98eda74a322efb9fbd85540856ae6d9 *dmd.stage3.o
```
Oops! I'll see if I can do a closer analysis as to the why, but my initial assumption is something like a timestamp is making the difference.

**Comment by MartinNowak** [comment]
As I already said (https://trello.com/c/4tefvo4i/51-add-three-stage-builds-for-ddmd), this doesn't even work for the C++ builds, b/c of timestamps and unique build ids.

**Comment by MartinNowak** [comment]
This also tests ddmd with the master druntime/phobos which is needed anyhow.

**Comment by ibuclaw** [comment]
> this doesn't even work for the C++ builds, b/c of timestamps and unique build ids.

Really?  GCC passes its own bootstrap checksums just fine, and I'd expect GDC so pass too when we make the switchover.

**Comment by ibuclaw** [comment]
It seems that the root of it is that DMD outputs undefined symbols in an non-deterministic way.
```
-11374460   .   _   D   1   1   T   y   p   e   I   n   f   o   _   x   k
+11374460   .   _   D   1   1   T   y   p   e   I   n   f   o   _   x   v
``` 
In any case, this doesn't seem to be something that will be looked at any time soon.

**Comment by ibuclaw** [comment]
Auto-merge toggled on

**Comment by MartinNowak** [comment]
> Really? GCC passes its own bootstrap checksums just fine, and I'd expect GDC so pass too when we make the switchover.

Mmh, seems to work with ld.bfd but not with ld.gold.

**Comment by MartinNowak** [comment]
> It seems that the root of it is that DMD outputs undefined symbols in an non-deterministic way.

See https://github.com/D-Programming-Language/dmd/blob/20771f89a16d9ab663b2964db8d49706cfa0e1dd/src/backend/cgen.c#L501.

**Comment by ibuclaw** [comment]
> See https://github.com/D-Programming-Language/dmd/blob/20771f89a16d9ab663b2964db8d49706cfa0e1dd/src/backend/cgen.c#L501.

And now they are identical!
```
ad6e17a9294784d9666e2708a8da6c04 *stage2/dmd.o
ad6e17a9294784d9666e2708a8da6c04 *stage3/dmd.o
```
However, I changed my script to use ``cmp`` instead, which takes care of the comparisons for us.  Is there any way to make ``FLARRAY`` exposed to the build script?

**Comment by ibuclaw** [comment]
FYI, https://gist.github.com/ibuclaw/31d438291888f1f0b4d0

**Comment by yebblies** [comment]
So ASLR is changing addresses, causing a different permutation of relocation order to be emitted?  We really should fix the backend to stop doing that.  I tried and failed a couple of years ago.

**Comment by MartinNowak** [comment]
> Is there any way to make FLARRAY exposed to the build script?

Maybe simply use Stringtable instead of the symbol addresses?

**Comment by MartinNowak** [comment]
See https://issues.dlang.org/show_bug.cgi?id=14976

**Pull request merged**